clippy_lints/src/indexing_slicing.rs

```diff
@@ -0,0 +1,201 @@
+///
+/// **Why is this bad?** Indexing and slicing can panic at runtime and there are
+/// safe alternatives.
+///
+/// **Known problems:** Hopefully none.
+///
+/// **Example:**
+/// ```rust
+/// let x = vec![0; 5];
+/// // Bad
+/// x[2];
+/// &x[2..100];
+/// &x[2..];
+/// &x[..100];
+///
+/// // Good
+/// x.get(2)
+/// x.get(2..100)
+/// x.get(2..)
+/// x.get(..100)
+/// ```
+declare_clippy_lint! {
+    pub INDEXING_SLICING,
+    pedantic,
+    "indexing/slicing usage"
+}
+
+#[derive(Copy, Clone)]
+pub struct IndexingSlicing;
+
+impl LintPass for IndexingSlicing {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(INDEXING_SLICING, OUT_OF_BOUNDS_INDEXING)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for IndexingSlicing {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        if let ExprIndex(ref array, ref index) = &expr.node {
+            match &index.node {
+                // Both ExprStruct and ExprPath require this approach's checks
+                // on the `range` returned by `higher::range(cx, index)`.
+                // ExprStruct handles &x[n..m], &x[n..] and &x[..n].
+                // ExprPath handles &x[..] and x[var]
+                ExprStruct(..) | ExprPath(..) => {
+                    if let Some(range) = higher::range(cx, index) {
+                        let ty = cx.tables.expr_ty(array);
+                        if let ty::TyArray(_, s) = ty.sty {
+                            let size: u128 = s.assert_usize(cx.tcx).unwrap().into();
+                            // Index is a constant range.
+                            if let Some((start, end)) = to_const_range(cx, range, size) {
+                                if start > size || end > size {
+                                    utils::span_lint(
+                                        cx,
+                                        OUT_OF_BOUNDS_INDEXING,
+                                        expr.span,
+                                        "range is out of bounds",
+                                    );
+                                } else {
+                                    // Range is in bounds, ok.
+                                    return;
+                                }
+                            }
+                        }
+
+                        let help_msg;
+                        match (range.start, range.end) {
+                            (None, Some(_)) => {
+                                help_msg = "Consider using `.get(..n)`or `.get_mut(..n)` instead";
+                            }
+                            (Some(_), None) => {
+                                help_msg = "Consider using `.get(n..)` or .get_mut(n..)` instead";
+                            }
+                            (Some(_), Some(_)) => {
+                                help_msg =
+                                    "Consider using `.get(n..m)` or `.get_mut(n..m)` instead";
+                            }
```

Comment:
You could prettify this a little bit: see [playground](https://play.rust-lang.org/?gist=49121799238a875ad62557b7b96da137&version=stable&mode=debug)